tests/integration_tests/tests/connect_info.rs

```diff
@@ -48,3 +48,80 @@ async fn getting_connect_info() {
 
     jh.await.unwrap();
 }
+
+#[cfg(unix)]
+pub mod unix {
+    use std::path::Path;
+
+    use futures_util::FutureExt;
+    use tokio::{
+        net::{UnixListener, UnixStream},
+        sync::oneshot,
+    };
+    use tokio_stream::wrappers::UnixListenerStream;
+    use tonic::{
+        transport::{server::UdsConnectInfo, Endpoint, Server, Uri},
+        Request, Response, Status,
+    };
+    use tower::service_fn;
+
+    use integration_tests::pb::{test_client, test_server, Input, Output};
+
+    struct Svc {}
+
+    #[tonic::async_trait]
+    impl test_server::Test for Svc {
+        async fn unary_call(&self, req: Request<Input>) -> Result<Response<Output>, Status> {
+            let conn_info = req.extensions().get::<UdsConnectInfo>().unwrap();
+
+            // Client-side unix sockets are unnamed.
+            assert!(req.remote_addr().is_none());
+            assert!(conn_info.peer_addr.as_ref().unwrap().is_unnamed());
+            // This should contain process credentials for the client socket.
+            assert!(conn_info.peer_cred.as_ref().is_some());
+
+            Ok(Response::new(Output {}))
+        }
+    }
+
+    #[tokio::test]
+    async fn getting_connect_info() {
+        let unix_socket_path = "/tmp/uds-integration-test";
+
+        // tokio's `UnixListener` does not cleanup the socket automatically - we need to manually
+        // remove the file at the end of the test.
+        tokio::fs::create_dir_all(Path::new(unix_socket_path).parent().unwrap())
+            .await
+            .unwrap();
+        let uds = UnixListener::bind(unix_socket_path).unwrap();
+        let uds_stream = UnixListenerStream::new(uds);
+
+        let service = test_server::TestServer::new(Svc {});
+        let (tx, rx) = oneshot::channel::<()>();
+
+        let jh = tokio::spawn(async move {
+            Server::builder()
+                .add_service(service)
+                .serve_with_incoming_shutdown(uds_stream, rx.map(drop))
+                .await
+                .unwrap();
+        });
+
+        let channel = Endpoint::try_from("http://[::]:50051")
+            .unwrap()
+            .connect_with_connector(service_fn(move |_: Uri| {
+                UnixStream::connect(unix_socket_path)
+            }))
+            .await
+            .unwrap();
```

Comment:
How easy would it be to make `Endpoint::try_from("unix:///tmp/uds-integration-test)` work without the fake url/custom connector? Probably fine to punt.

Reply:
Seems like this would get a bit hairy as implementing something in `Endpoint` would bubble back up though `tonic::transport::Channel` and `tonic::transport::service::Connection` and we'd have to find some way to handle an optional URI inside an `Endpoint` through those (and update tests, etc). Seems like more trouble than it's worth considering this only saves consumers 1 LOC